### PR TITLE
Implement VMC_SR with various run modes.

### PR DESCRIPTION
**Pull request by PhilipVinc**
Proper implementation of a `VMC_SR` driver.
I would prefer to call it `VMC_NGD`, but probably @gcarleo would disagree.

Follows my proposal from https://github.com/netket/netket/pull/1987#issuecomment-2631450927
Used as 
```python
gs = nkx.driver.VMC_SR(
    ha, op, variational_state=vs, diag_shift=0.001, use_ntk=True, on_the_fly=True
)
```
Argument names are:

- `use_ntk=True/False/None` for using the kernel trick/NTK/minSR or not. None will automatically switch to it when there are more parameters then samples
- `on_the_fly=True/False/None` for using the lazy computation of the NTK/QGT matrix based on jvp/vjps instead of the jacobian. 
- `mode="real"/"complex"` to switch between real and complex modes as usual
- `momentum=0.XX/None` from PR #1708  

Quick comments about argument names are welcome.
This deprecates the previous `VMC_SRt`.

- The code is modular in order for it to be used for other drivers in the future. 
- It is based on some WIP distributed primitives to ensure that it works under sharding/MPI,  battle tested over the last few months internally. I think we caught most bugs.
- It vendors neural_tangents, and adds a script in `ext/` to update the vendored copy if necessary in the future. The veneered version preserves the original license file as good behaviour requires.

Credit for the content of this PR go largely to @lgravina1997 who wrote the modular version and the too many tests, but also to @AmedSho97 who got the distributed working, @rajnutakki for catching a long list of bugs and edge cases in the sharding.

Credits also to @wdphy16 who wrote the original SPRING/Momentum implementation, and @ilviteritti / @riccardo-rende for parts of this implementation.

### CODE REVIEW

File: netket/experimental/driver/vmc_ngd.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+    .. note:: Citation
+
+        - Stochastic Reconfiguration was originally introduced in the QMC field by `Sorella <https://arxiv.org/abs/cond-mat/9803107>`_.
+          The method was later shown to be equivalent to the Natural Gradient Descent method introduced by
+          `Amari <https://ieeexplore.ieee.org/abstract/document/6790500/>`_ in the context of neural network training.
+
+        - The *kernel tricK* which makes NGD/SR feasible in the large-parameter count limit was originally introduced
+          to the field of NQS by `Chen & Heyl <https://arxiv.org/abs/2302.01941>`_ who called it `minSR`. However,
+          `Rende & Al <https://arxiv.org/abs/2310.05715>`_ delivered a more straightforward derivation in terms of
+          the Kernel trick. In order to not pick a side, we will however mention that the kernel trick for Natural Gradient Descent
```
Comment (gcarleo) [inline netket/experimental/driver/vmc_ngd.py:121]: I would avoid this jovial remark and just add the citation (again mentioning it was for the real NGD)

File: netket/experimental/driver/vmc_ngd.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+          was first mentioned by `Ren & Goldfarb <https://arxiv.org/abs/1906.02353>`_.
+
+        - The two derivations (Chen and Rende) are equivalent when assuming the wavefunction has no nodes. (As to when the
+          wavefunction has nodes, it's still a subject of friendly yet heated discussions every time i meet Markus Sr.)
```
Comment (gcarleo) [inline netket/experimental/driver/vmc_ngd.py:125]: I would also remove this, it is not clear if it Netket itself that is meeting with Markus Heyl? also since it's not clear what the issue is and if it's relevant

**Review by gcarleo — CHANGES_REQUESTED** [review]

**Comment by gcarleo** [comment]
Let me add that the original implementation of the NTK Sr was provided by @riccardo-rende and @llviteritti, and that @inailuig did the ground work for the distributed implementation used here

**Comment by PhilipVinc** [comment]
It's okay to keep it as SR. 
I don't like the distinction of NGD \neq qNGD. Yes they are two different things, but it's largely 'us' stating that. 
The literature of NGD or Riemannian Flows always discusses the 'natural metric'. It just so happens that most of those works focus on probability distribution and therefore on the Fisher Matrix/KL.

The insight of looking at the right metric/space is the good one, and there's a lot of relevant literature and my comments are to point people towards that literature.

File: netket/experimental/driver/vmc_ngd.py (hunk 2 above)
Comment (PhilipVinc) [reply netket/experimental/driver/vmc_ngd.py]: Yeah, I forgot to cleanup the comments before moving the code around.

However, I doubt there is any doubt on who is making this commentary...

File: netket/experimental/driver/vmc_ngd.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        computationally efficient for networks that reuse the parameters many times for every forward pass
+        (convolutions, attention layers, but not dense layers...) and generally uses less memroy.
+
+        However, the on the fly implementation relies on some JAX compiler behaaviour, so it might at times
+        have worse performance. We suggest you check on your specific model.
+
+    .. note:: Citation
+
+        - Stochastic Reconfiguration was originally introduced in the QMC field by `Sorella <https://arxiv.org/abs/cond-mat/9803107>`_.
+          The method was later shown to be equivalent to the Natural Gradient Descent method introduced by
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/experimental/driver/vmc_ngd.py]: I changed it here. 
I still keep several references to NGD literature.
There are not a lot of methodological studies on SR, and I want people that want to dig into something in a relevant direction.

**Comment by gcarleo** [comment]
I believe in this paper https://quantum-journal.org/papers/q-2020-05-25-269/ we made the connection clear, clarifying also the difference with the Fisher. It is true they are similar, I am not denying it, I am just saying we should nuance the statement "it is equivalent to NGD" I would rather say "it is similar in spirit to NGD" something like that ?

File: netket/_src/ngd/sr_srt_common.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    solver_fn: Callable[[Array, Array], Array],
+    mode: str,
+    proj_reg: Optional[Union[float, Array]] = None,
+    momentum: Optional[Union[float, Array]] = None,
+    old_updates: Optional[PyTree] = None,
+    chunk_size: Optional[int] = None,
+    use_ntk: bool = False,
+):
+    r"""
+    Compute the Natural gradient update for the model specified by `log_psi({parameters, model_state}, samples)`
```
Comment (gcarleo) [inline netket/_src/ngd/sr_srt_common.py:92]: also this

**Comment by llviteritti** [comment]
All the people in the community call this method SR. In my opinion, it's useful to keep this name to avoid creating further confusion.

**Review by ggoldsh — COMMENTED** [review]
I stumbled on this PR and am excited to see that SPRING will be incorporated with these updates to NetKet's SR implementation :)

I looked through the SPRING part and left a couple comments to provide more explanation on the theory behind SPRING and on how SPRING relates to traditional momentum schemes. Hope this is helpful for you guys and for any NetKet users who end up using this functionality!

File: netket/experimental/driver/vmc_sr.py [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+    --------
+    When `momentum` is used, this driver implements the SPRING optimizer in
+    `G.Goldshlager et Al. <https://arxiv.org/abs/2401.10190>`_
+    to accumulate previous updates for better approximation of the exact SR with
+    no significant performance penalty.
+
+    `momentum` is a number between [0,1] that specifies the damping factor of the previous updates, and
+    works similarly to the beta parameter of ADAM. Therefore a value of `momentum=0.9` corresponds to a
+    moderate damping factor of 0.1, and means that if subsequent moves are close, the steps might be
+    amplified by at most a factor of 10. A value of `momentum=0.99` corresponds to a damping factor of 0.01,
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/experimental/driver/vmc_sr.py:126]: Thanks, that's interesting.

I remember reading your first paper a few years ago and having the feeling that what you were doing was indeed some sort of variance reduction but the message of the paper felt to me more about momentum (though I might remember something erroneously).

If you have some more precise statements to add here feel free to propose a 'new' blurb to replace what I wrote.

Also, if you think that momentum is not a good keyword (seems not) then do propose a new one!

[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/experimental/driver/vmc_sr.py:126]: I think what you are saying above is very relevant and it would be worthwhile to mention it.

By the way, if you had a few simple toy cases that you can use to cook up a 'simple explanation of what SPRING does, it would be nice if you could conjure up a notebook to show that off and we could include it somewhere on the documentation.

[the remark this answers is not in the document]
Comment (ggoldsh) [reply netket/experimental/driver/vmc_sr.py:126]: ```I remember reading your first paper a few years ago and having the feeling that what you were doing was indeed some sort of variance reduction but the message of the paper felt to me more about momentum (though I might remember something erroneously).```

Indeed the messaging is a bit confusing and I am still somewhat ambivalent about how to describe this. Overall I still feel that it is more clear to label it as momentum because momentum is a known concept and it does appear on the surface to be similar to a momentum method. I would suggest the following description:

----
"momentum" is a number between `[0,1]` that specifies the damping factor of the previous updates and
 works somewhat similarly to the beta parameter of ADAM. The difference is that rather than simply adding the damped previous update to the new update, SPRING uses the damped previous update to fill in the components of the SR direction that are not sampled by the current batch of walkers, resulting in a more accurate and less noisy estimate. Since SPRING only uses the previous update to fill in directions that are orthogonal to the current one, the maximum amplification of the step size in SPRING is `1/sqrt(1-momentum^2)` rather than `1/(1-momentum)`. Thus the amplification is at most a factor of `2.3` when `momentum=0.9` and `7.1` when `momentum=0.99`. 

Values that empirically work are around `0.8.`

Some progress has been made on theoretically analyzing this parameter, in particular Section 3 of https://arxiv.org/pdf/2411.19877 demonstrates (albeit in a significantly simplified linear least-squares setting) that SPRING can be interpreted as iteratively estimating a regularized SR direction, with the amount of regularization proportional to the value of `1-momentum`. Additional insights regarding the behavior of some SPRING-like algorithms, albeit still in the linear least-squares setting, are presented in https://arxiv.org/pdf/2502.00882.

Nonetheless, further contributions along these lines, particularly concerning ways to understand the convergence of SPRING in the full VMC setting, are welcome.

----

Of course, feel free to touch this up as you have a better sense than I do of what will make sense and be most helpful to users who are not already familiar with these algorithms.

[the remark this answers is not in the document]
Comment (ggoldsh) [reply netket/experimental/driver/vmc_sr.py:126]: Regarding the notebook, that's a nice idea and I have some ideas of how to go about it. I'll  try to cook something up when I have some time.

**Comment by jwnys** [comment]
Also, why do we need this external package? It's quite messy, but I think we don't need much of it... Something like the following does the same as what we need, unless I'm misundertanding things.
```python
import jax
import jax.numpy as jnp
from functools import partial

def empirical_ntk_fn(apply_fun):
    """ Return jacobian contraction function, but reshift the inputs to my convention."""
    def _jacobian_contraction(x, y, parameters): # structure netket expects
        return jacobian_contraction_real(apply_fun, parameters, x, y=y) # my form
    return _jacobian_contraction 
    
def jacobian_contraction_real(
    apply_fun,
    parameters,
    x,
    y=None,
):
    """
    Computes the jacobian contraction of a function with respect to its parameters, for real-valued inputs and parameters only.
    """
    if x.ndim >= 3:
        x = jax.lax.collapse(x, 0, 2)
    if y is not None and y.ndim >= 3:
        y = jax.lax.collapse(y, 0, 2)
    
    # Compute jacobian on x batch
    jx = jacobian(apply_fun, parameters, x)
    if y is None:
        jy = jx 
    else:
        jy = jacobian(apply_fun, parameters, y)
    return sum_and_contract(jx, jy)
    
def jacobian(fn, params, x):
    # fn(x) # can be (N_mc,2) or (N_mc,), we always make it to (N_mc, (1 or 2))
    _, hilbert_size = x.shape
    def _fn_single(params, x):
        fn_x = fn(params, x.reshape(1, hilbert_size))
        fn_x = fn_x.reshape(-1) # always (1 or 2)
        assert fn_x.ndim == 1
        assert fn_x.size in (1, 2)
        return fn_x
    return jax.vmap(jax.jacobian(_fn_single), in_axes=(None, 0))(params, x)
    
def sum_and_contract(jx, jy):    
    """ For each parameter block, sum over the parameters axis. """
    block_ntk = jax.tree.map(contract_block, jx, jy) 
    return jax.tree.reduce(jnp.add, block_ntk)
        
@partial(jax.vmap, in_axes=(0, None), out_axes=0) # vmap x samples
@partial(jax.vmap, in_axes=(None, 0), out_axes=0) # vmap y samples
def contract_block(jx, jy):
    # in every block: jx = (1 or 2, parameter block shape)
    assert jx.shape[0] == jy.shape[0] # same output shape
    output_size = jx.shape[0]
    assert output_size in (1, 2)
    # flatten parameter block
    jx = jx.reshape(jx.shape[0], -1)
    jy = jy.reshape(jy.shape[0], -1)    
    assert jx.shape[-1] == jy.shape[-1] # same number of parameters in the block
    jcontr = jx @ jy.T # sum over the parameter dimension, BUT: in the other version they have a 1/size???
    assert jcontr.shape == (output_size, output_size)
    if output_size == 1:
        # netket expects output to be (N_mc, N_mc)
        jcontr = jnp.squeeze(jcontr, axis=(-2, -1))
    return jcontr

```

**Comment by PhilipVinc** [comment]
> Also, why do we need this external package? It's quite messy, but I think we don't need much of it... Something like the following does the same as what we need, unless I'm misundertanding things.

Lack of time, because I had the stuff to vendor everything at hand.
But I agree that implementing only the 2 functions that are needed would be the best solution.
Also because neural tangents is essentially unmaintained and throws errors in jax 0.6.

> This looks really nice, and I'm giving it a go. I'm getting errors though related to the shard_map I imagine (in jax 0.5.3), suggesting the shard_map returns None. Any idea what's happened with shard map?

It's a series of bugs in jax that are fixed in jax 0.6, but 0.6 completely breaks netket in several places because of changes to shard map.

**Comment by PhilipVinc** [comment]
@gcarleo if you are playing with codex you should try to make him find the bug in this PR.
There's an hidden bug that emerges only when running under MPI (not when running with sharding) which is why I've never finished this.
I tried many times to find it and never could, that's why I never finished this.
It's why the tests fail.

Running `mpirun -np 2 pytest -n0 test/driver/test_vmc_ngd.py` (with mpi4jax installed) should fail. I've checked the code many times and never found the bug.

(On a side note, this made me think about dropping MPI for good.)

**Comment by gcarleo** [comment]
yeah I'm note sure it works with mpi, I'll try ...

**Comment by gcarleo** [comment]
can you quickly create this as a netket branch maybe? otherwise codex cannot access it on your github...

**Comment by PhilipVinc** [comment]
Most comments were addressed. 
The misteriously failing tests on MPI disappear now that we drop MPI, and since we're fixing all the bugs necessary to go to jax 0.7, better do it on the new code.
